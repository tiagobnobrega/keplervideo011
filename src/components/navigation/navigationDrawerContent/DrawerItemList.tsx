import {
  HWEvent,
  TVFocusGuideView,
  useTVEventHandler,
} from '@amzn/react-native-kepler';
import {
  DrawerDescriptorMap,
  DrawerNavigationHelpers,
} from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import {
  DrawerNavigationState,
  ParamListBase,
} from '@amzn/react-navigation__native';
import {
  NavigationState,
  PartialState,
  Route,
} from '@amzn/react-navigation__routers/lib/typescript/src/types';
import { isEqual } from 'lodash';
import React, { useEffect, useState } from 'react';
import { findNodeHandle, StyleSheet, TouchableOpacity } from 'react-native';
import { useSelector } from 'react-redux';
import { DPADEventType, EVENT_KEY_DOWN } from '../../../constants';
import { focusSelectors } from '../../../store/focus/focusSlice';
import { recordCustomEvent } from '../../../utils/helperNewRelic';
import { Screens } from '../types';
import DrawerItem from './DrawerItem';

type NavigationRoute<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList,
> = Route<Extract<RouteName, string>, ParamList[RouteName]> & {
  state?: NavigationState | PartialState<NavigationState>;
};

export interface DrawerItemListProps {
  state: DrawerNavigationState<ParamListBase>;
  navigation: DrawerNavigationHelpers;
  descriptors: DrawerDescriptorMap;
  isDrawerInFocus: boolean;
  onDrawerListFocus: () => void;
  onDrawerListBlur: () => void;
  drawerWidth: number;
  testID?: string;
}

const NOT_FOCUSED_ITEM = -1;

const DrawerItemList = ({
  state,
  navigation,
  descriptors,
  onDrawerListFocus,
  onDrawerListBlur,
  isDrawerInFocus,
  drawerWidth,
  testID,
}: DrawerItemListProps) => {
  const [rootNodeHandle, setRootNodeHandle] = useState<number | undefined>(
    undefined,
  );

  const [selectedItem, setSelectedItem] = useState(0);
  const [focusedItem, setFocusedItem] = useState(NOT_FOCUSED_ITEM);
  const currentFocusScreen = useSelector(focusSelectors.getCurrentFocus);

  useTVEventHandler((evt: HWEvent) => {
    if (!isDrawerInFocus) {
      return;
    }

    if (evt?.eventKeyAction !== EVENT_KEY_DOWN) {
      return;
    }

    if (evt?.eventType === DPADEventType.SELECT) {
      onDrawerListBlur();
    }
    if (evt.eventType === DPADEventType.UP && focusedItem > 0) {
      setFocusedItem(focusedItem - 1);
    } else if (
      evt.eventType === DPADEventType.DOWN &&
      focusedItem < state.routes.length - 1
    ) {
      setFocusedItem(focusedItem + 1);
    } else if (evt.eventType === DPADEventType.SELECT) {
      const { name } = state.routes[focusedItem];
      onItemPress(name);
    }
  });

  useEffect(() => {
    if (currentFocusScreen === Screens.DEFAULT_SCREEN) {
      setFocusedItem(0);
      setSelectedItem(0);
    }
  }, [currentFocusScreen]);

  useEffect(() => {
    if (!isDrawerInFocus) {
      setFocusedItem(NOT_FOCUSED_ITEM);
    }
  }, [isDrawerInFocus]);

  const onItemPress = (routeName: string) => {
    const index: number = state.routes.findIndex(r => r.name === routeName);
    setFocusedItem(index);
    setSelectedItem(index);
    navigation.navigate(routeName);
    recordCustomEvent('Navigated: ', {
      name: routeName,
    });
  };

  const onFocus = () => {
    setFocusedItem(selectedItem);
    onDrawerListFocus();
  };

  const onBlur = () => {
    setFocusedItem(NOT_FOCUSED_ITEM);
    onDrawerListBlur();
  };

  return (
    <TVFocusGuideView style={styles.listContainer} trapFocusDown>
      <TouchableOpacity
        activeOpacity={1}
        ref={component => {
          const nodeHandle = findNodeHandle(component);
          setRootNodeHandle(nodeHandle ? nodeHandle : undefined);
        }}
        style={styles.listContainer}
        nextFocusDown={rootNodeHandle}
        nextFocusUp={rootNodeHandle}
        onFocus={onFocus}
        onBlur={onBlur}
        testID={testID}>
        {state.routes.map(
          (route: NavigationRoute<ParamListBase, string>, index: number) => {
            const { title, drawerLabel, drawerIcon } =
              descriptors[route.key].options;
            return (
              <DrawerItem
                key={route.key}
                label={title || (drawerLabel as string) || route.name}
                icon={drawerIcon as any}
                isSelected={selectedItem === index}
                isFocused={focusedItem === index}
                drawerWidth={drawerWidth}
                isDrawerInFocus={isDrawerInFocus}
                onPress={onItemPress}
                routeName={route.name}
              />
            );
          },
        )}
      </TouchableOpacity>
    </TVFocusGuideView>
  );
};

const drawerItemDataUnchanged = (
  prevProps: DrawerItemListProps,
  nextProps: DrawerItemListProps,
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(DrawerItemList, drawerItemDataUnchanged);

const styles = StyleSheet.create({
  listContainer: {
    flex: 1,
    justifyContent: 'center',
  },
});
