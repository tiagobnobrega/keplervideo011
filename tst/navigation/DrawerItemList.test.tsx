import {
  DrawerDescriptorMap,
  DrawerNavigationHelpers,
} from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import {
  DrawerNavigationState,
  ParamListBase,
} from '@amzn/react-navigation__routers';
import { render } from '@testing-library/react-native';
import { isEqual } from 'lodash';
import React from 'react';
import DrawerItemList, {
  DrawerItemListProps,
} from '../../src/components/navigation/navigationDrawerContent/DrawerItemList';

const mockState = {
  index: 0,
  routes: [
    { key: 'home', name: 'Home' },
    { key: 'search', name: 'Search' },
    { key: 'settings', name: 'Settings' },
  ],
} as unknown as DrawerNavigationState<ParamListBase>;
const mockDescriptors = {
  home: {
    options: {
      title: 'Home',
    },
  },
  search: {
    options: {
      title: 'Search',
    },
  },
  settings: {
    options: {
      title: 'Settings',
    },
  },
} as unknown as DrawerDescriptorMap;
const mockNavigation = {
  navigate: jest.fn(),
  dispatch: jest.fn(),
} as unknown as DrawerNavigationHelpers;

const onDrawerListFocusMock = jest.fn();
const onDrawerListBlurMock = jest.fn();
const mockNavigate = jest.fn();
const mockSetFocusedItem = jest.fn();
const mockSetSelectedItem = jest.fn();

const mockProps = {
  state: mockState,
  navigation: mockNavigation,
  descriptors: mockDescriptors,
  isDrawerInFocus: false,
  onDrawerListFocus: onDrawerListFocusMock,
  onDrawerListBlur: onDrawerListBlurMock,
  drawerWidth: 0,
};
const renderDrawerItemList = (props?: Partial<DrawerItemListProps>) => {
  return (
    <DrawerItemList
      state={mockState}
      navigation={mockNavigation}
      descriptors={mockDescriptors}
      isDrawerInFocus={false}
      onDrawerListFocus={onDrawerListFocusMock}
      onDrawerListBlur={onDrawerListBlurMock}
      drawerWidth={0}
      {...props}
    />
  );
};

// Define the onItemPress function
const onItemPress = (routeName: string) => {
  const index = mockState.routes.findIndex(r => r.name === routeName);
  mockSetFocusedItem(index);
  mockSetSelectedItem(index);
  mockNavigate(routeName);
};

describe('DrawerItemList', () => {
  it('renders correctly', () => {
    const component = render(renderDrawerItemList());
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with', () => {
    const component = render(renderDrawerItemList());
    expect(component).toMatchSnapshot();
  });

  it('should set focused item, set selected item, and navigate to the route', () => {
    // Act: Call the onItemPress function with a route name
    const routeName = 'Profile';
    onItemPress(routeName);

    // Assert: Check if the correct functions were called with expected values
    const expectedIndex = mockState.routes.findIndex(r => r.name === routeName);

    expect(mockSetFocusedItem).toHaveBeenCalledWith(expectedIndex);
    expect(mockSetSelectedItem).toHaveBeenCalledWith(expectedIndex);
    expect(mockNavigate).toHaveBeenCalledWith(routeName);
  });

  it('re-renders when props are not equal', () => {
    const { rerender } = render(<DrawerItemList {...mockProps} />);
    rerender(<DrawerItemList {...mockProps} />);
    expect(jest.mocked(isEqual)).toHaveBeenCalled();
  });
});
