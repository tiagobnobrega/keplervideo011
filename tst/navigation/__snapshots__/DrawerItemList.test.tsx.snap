// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerItemList renders correctly 1`] = `null`;

exports[`DrawerItemList renders correctly with 1`] = `null`;
